### DISCUSSION TITLE
Restart dev server on .env.* file changes

### DISCUSSION THREAD

**Discussion by ArianHamdi** [vercel/next.js#59558] [Ideas]
### Goals

1. Automatically restart the Next.js dev server when changes are detected in ```.env.*``` files.
2. Enhance developer experience by ensuring that configuration changes are reflected without manual server restarts.

### Background

Currently, the Next.js dev server restarts when changes are detected in ```next.config.js```. However, when ```.env.*``` files are modified, the dev server does not automatically restart. This can lead to a suboptimal developer experience as manual restarts are required to apply the updated configuration.

### Proposal

Implement a mechanism in the Next.js dev server to detect changes in ```.env.*``` files and trigger an automatic restart when such changes are detected. This ensures that any modifications to environment variables are immediately reflected in the running development server without the need for manual intervention.

**Comment by pybern** [comment]
I am on `Next.js 14` and it reloads my `.env` file when its updated.

![image](https://github.com/vercel/next.js/assets/4676736/29036f98-e966-4ebf-ab1e-f90ed795bd4a)

**Reply by ArianHamdi** [reply]
It reloads the `.env.*`, so if you log environment variables, you will see new values. However, it doesn't restart the dev server. Therefore, if you used `.env` files inside `next.config.js`, changing them won't affect the server.

**Comment by FindingJohnny** [comment]
Looking for this exact feature.

Use Case:
We have a brand specific themes that we have built using SCSS Modules that we are copy to a single location for reference from the code base at build time.

Import Example:
```scss
@import "@/scss/themes/brands/colors"
```

Config Example Below:
```javascript
const nextConfig = {
  reactStrictMode: true,
  webpack: (config, options) => {

    config.plugins.push(
      new CopyPlugin({
        patterns: [
          {
            from: `src/scss/themes/${process.env.NEXT_PUBLIC_BRAND}`,
            to: '../src/scss/themes/brand',
            force: true,
          },
        ],
      })
    );

    return config;
  },
};

module.exports = nextConfig;
```

**Comment by yinebebt** [comment]
Does this discussion open yet?

**Comment by ArianHamdi** [comment]
Yeap

**Reply by yinebebt** [reply]
ok, we will check it with my teams.

**Comment by ArianHamdi** [comment]
Is there any progress on this? it can be a bit problematic when `next.config.js` rely on environment variables.

**Comment by trishantpahwa** [comment]
Why not allow the custom file names to be listed in next.config.js that restart the server on any change in them?

**Comment by bmight** [comment]
Hoping to revive this discussion. I have a similar issue when using mongoose. Mongoose recommends preventing model registration issues by first checking if the model is already registered. 

```
// models/User.js
import mongoose from 'mongoose';

const UserSchema = new mongoose.Schema({
  name: String,
  email: { type: String, required: true }
}, { timestamps: true });

export default mongoose.models.User || mongoose.model('User', UserSchema);
```
[Using Mongoose With Next.js - Model Registration](https://mongoosejs.com/docs/nextjs.html#model-registration)

This is great for 99% of changes because it uses the cached model. The downside is when you change a model you have to manually restart the server. It would be nice to add custom glob file paths to config or dev cli flag that forces the server to restart like `next.config.ts` does.

**Comment by yakubka** [comment]
Next.js has had file-system watching for .env.* files since version 14. When the dev server detects a change in .env, .env.local, .env.development, or .env.development.local it shows a message in the terminal and hot-reloads the environment for new requests without a full manual restart.

If you are not seeing this behavior there are a few common reasons.

The first is that you are on a version older than 14. Upgrading to the latest stable release is the easiest fix.

The second is that your .env file is outside the project root. The watcher only covers files at the root level next to next.config.js.

The third is that you are changing a variable that is baked into the bundle at build time, specifically any NEXT_PUBLIC_ variable. Those are inlined by the compiler during the initial build step, so a hot reload of the env file is not enough. You need a full restart for NEXT_PUBLIC_ changes because the compiled output must be regenerated. Server-side variables like DATABASE_URL or API_KEY work fine with the hot reload because they are read at request time.

For the case where next.config.js reads env vars at startup (like the SCSS theme scenario above), the file watcher restarts the entire dev server when next.config.js itself changes, but it does not do a full restart when only .env changes. A pattern that works there is to move the env-dependent config into a dynamic lookup so it is evaluated per request rather than at startup time.

**Comment by vimal-gandhi-dev** [comment]
One workaround that's helped me is moving values that need to change during development out of next.config.js and reading them directly with process.env at runtime where possible (e.g. inside server components, API routes, or getServerSideProps). Since Next.js 14+ already watches .env* files and reloads env values for server-side code read at request time, this avoids needing a full restart for those values.
The catch is anything actually evaluated inside next.config.js (like config-time imports or webpack plugin options) still requires a restart, since that file only runs once at server startup. And this trick only helps for server-only vars — anything prefixed NEXT_PUBLIC_ gets inlined into the client bundle at build time, so those still need a full rebuild/restart regardless of where you read them from.

**Reply by yakubka** [reply]
That is the right pattern for most cases. The important distinction is config-time vs request-time: anything inside next.config.js is evaluated once when the dev server starts, so environment variables read there are frozen at startup. Anything evaluated inside a server component, route handler, or getServerSideProps reads process.env fresh on each request, which is what the Next.js hot-reload for .env files updates. Moving toward request-time reads wherever possible is the least-friction path to avoiding manual restarts.
